### DISCUSSION TITLE
Custom Casts to allow RAW Values

### DISCUSSION THREAD

**Discussion by ojsoft** [laravel/framework#47395] [Ideas] [answered]
I do have Point Geometry in my database table. It is currently not possible to have the point like

'Point(x,y)' as the attribute in the model and to save this to the database, because it handles the value as a string and does this on save:
`update model SET location = 'Point(x,y)' where ...`

What I need is this or MySql just does not accept it:
`update model SET location = Point(x,y) where ...`

Currently I override the save() method to get a workaround:

`
public function save(array $options = [])
{
      $location = $this->attributes['location'];
      $this->attributes['location'] = \DB::raw($location);
      $return = parent::save($options);
      $this->attributes['location'] = $location;
      return $return;
}
`

Well this is a bit hacky in my opinion. Of course I could save a \DB::raw Object in attributes['location'], but this is quiet nasty when reading it out again. There is just no nice solution to this problem with current Custom Casts. First try was to have a mutator and a Cast at once, letting the Cast convert to /DB::raw . Took me an hour to figure out why the Cast was not called on save when there was a Mutator ( function getLocationAttribute() ) too.

Therefore I request a Custom Cast Option to make the Cast result to be raw just on save. This is where it logically belongs to. If I save a number, the number will be value = 3 and not value = '3'. I need the same for my own attributes, even if PHP can only store them as text.

If someone has a better cleaner idea, I am open to learn something.

**Comment by monkeyphysics** [comment, marked as answer]
Does this not work?

`App\Casts\Point.php`

```php
<?php

namespace App\Casts;

use Illuminate\Support\Facades\DB;
use Illuminate\Contracts\Database\Eloquent\CastsAttributes;

class Point implements CastsAttributes
{
    public function get($model, $key, $value, $attributes)
    {
        // https://stackoverflow.com/questions/37467050/convert-mysqls-point-to-text-in-php
        return unpack('x4/corder/Ltype/dlng/dlat', $value);
    }

    public function set($model, $key, $value, $attributes)
    {
        return DB::raw(sprintf('POINT(%f, %f)', $value['lng'], $value['lat']));
    }
}
```

**Reply by Cepairda** [reply]
When I use the code above I get an error

```
$order = Order::create([
            'user_id' => $request->user_id,
            'pickup_location' => [$request->pickup_lat, $request->pickup_lng], 
            'dropoff_location' => [$request->dropoff_lat, $request->dropoff_lng],
            'status' => 'pending',
        ]);
        
echo $order->pickup_location;
// Get Error
// Object of class Illuminate\Database\Query\Expression could not be converted to string        
```

**Reply by monkeyphysics** [reply]
That error is not related to the Cast..?

**Comment by ojsoft** [comment]
Interesting. I will give it a try. Many thanks.

**Reply by ojsoft** [reply]
Looked at my Cast Model now, yes the solution is obvious. Now I feel a bit stupid.

**Reply by monkeyphysics** [reply]
No worries, I needed this myself yesterday and was actually helped by your comment and the stackoverflow I linked to in the code. Using `DB::raw` as a return value makes Casts pretty powerful.

Was a little bit surprised there's no 3rd party Casts for all the spatial MySQL fields, but this is a start.

Last thing: if you are actually going to use this, you might want to implement a `Point` object and return that, because
`[ 'lat' => ... , 'lng' => ... ]` is undesired micro architecture.

Created as a gist: https://gist.github.com/monkeyphysics/8fa32b4bd79e185bc0a42b6b2ecab427
